### PR TITLE
Memoize expensive code

### PR DESCRIPTION
**Pull request by andycroll**

### CODE REVIEW

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  end
+end
+```
+
+
+## Instead...
+
+```ruby
+class OldTimeySweetShop
+`
```
Comment (nodunayo) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]: Stray backtick...!

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+shop.sweet_count
+#=> 4000003 # this one will be super quick!
+```
+
+
+## Why?
+
+Primarily this is a performance optimization. The first call of the method stores the result of an expensive operation in an instance variable that is then returned.
+
+When the method is called again the stored instance value is used and the expensive calculation is avoided.
```
Comment (nodunayo) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]: I feel like you've just repeated what you said in the intro...?

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+In typical Rails applications, I've found this patten to be most useful to optimize expensive database calls and to temporarily cache API requests to external services.
```
Comment (nodunayo) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]: Maybe all you need in this section is this part?

Comment (nodunayo) [reply _posts/2021/2021-01-25-memoize-expensive-code.md]: *pattern

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+
+## Why not?
+
+A piece of code can only be memoized _only if_ calling the code again would have the same output as replacing that function call with its return value.
```
Comment (nodunayo) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]: You repeat 'only' twice in the first sentence. I think you can remove the first one.

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+If you're looking at method calls with parameters, you can create a memoized lookup table, but it's more complex than the examples in this article.
```
Comment (nodunayo) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]: *but _that's_ more complex...

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+Memoizing results of repeated database queries isn't strictly necessary as Rails does it's own caching at the SQL level.
```
Comment (nodunayo) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]: *its

File: _posts/2021/2021-01-25-memoize-expensive-code.md (hunk 1 above)
Comment (andycroll) [reply _posts/2021/2021-01-25-memoize-expensive-code.md]:
```suggestion
```

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+
+## Instead of ...
+
+...repeating potentially expensive calculations:
+
+```ruby
+class OldTimeySweetShop
+
+  def average_sweets_per_jar
```
Comment (andycroll) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]:
```suggestion
  def average_sweets_per_jar
```

File: _posts/2021/2021-01-25-memoize-expensive-code.md (hunk 3 above)
Comment (andycroll) [reply _posts/2021/2021-01-25-memoize-expensive-code.md]:
```suggestion
In typical Rails applications, I've found this pattern to be most useful to optimize expensive database calls and to temporarily cache API requests to external services.
```

File: _posts/2021/2021-01-25-memoize-expensive-code.md (hunk 4 above)
Comment (andycroll) [reply _posts/2021/2021-01-25-memoize-expensive-code.md]:
```suggestion
A piece of code can be memoized _only if_ calling the code again would have the same output as replacing that function call with its return value.
```

File: _posts/2021/2021-01-25-memoize-expensive-code.md (hunk 5 above)
Comment (andycroll) [reply _posts/2021/2021-01-25-memoize-expensive-code.md]:
```suggestion
If you're looking at method calls with parameters, you can create a memoized lookup table, but that's more complex than the examples in this article.
```

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+In typical Rails applications, I've found this pattern to be most useful to optimize expensive database calls and to temporarily cache API requests to external services.
+
+
+## Why not?
+
+A piece of code can be memoized _only if_ calling the code again would have the same output as replacing that function call with its return value.
+
+If you're looking at method calls with parameters, you can create a memoized lookup table, but that's more complex than the examples in this article.
+
+Memoizing results of repeated database queries isn't strictly necessary as Rails does it's own caching at the SQL level.
```
Comment (andycroll) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]:
```suggestion
Memoizing results of repeated database queries isn't strictly necessary as Rails does its own caching at the SQL level.
```

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+shop.sweet_count
+#=> 4000003 # this one will be super quick!
+```
+
+
+## Why?
+
+Primarily this is a performance optimization. The first call of the method stores the result of an expensive operation in an instance variable that is then returned.
+
+When the method is called again the stored instance value is used and the expensive calculation is avoided.
```
Comment (andycroll) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]:
```suggestion
Primarily this is a performance optimization that has only a small impact on the readability of your code.
```

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+layout: article
+category: ruby
+image:
+  base: '2021/memoize-expensive-code'
+  alt: "Red Floppy Disk"
+  credit:  Fredy Jacob
+  source: "https://unsplash.com/photos/t0SlmanfFcg"
+---
+
+Memoization is a performance optimization where the result of a slow or non-performant piece of code is stored and when the expensive code is called again, the stored value is returned.
```
Comment (andycroll) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]:
```suggestion
Memoization is a performance optimization where the result of a slow or non-performant piece of code is temporarily stored and when the expensive code is called again, the stored value is returned.
```

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+image:
+  base: '2021/memoize-expensive-code'
+  alt: "Red Floppy Disk"
+  credit:  Fredy Jacob
+  source: "https://unsplash.com/photos/t0SlmanfFcg"
+---
+
+Memoization is a performance optimization where the result of a slow or non-performant piece of code is temporarily stored and when the expensive code is called again, the stored value is returned.
+
+It is a specific form of caching that is relatively elegant to read and implement in Ruby.
```
Comment (andycroll) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]:
```suggestion
It is a specific form of caching that is relatively elegant to read and implement in Ruby due to the use of the `||=` (or equals) operator.
```

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
+---
+title: "Memoize Expensive Code"
+description: ""
```
Comment (andycroll) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]:
```suggestion
description: "The ||= operator can provide significant performance improvements"
```

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+      jar.count_each_sweet_by_hand
+    end
+  end
+end
+
+shop = OldTimeySweetShop.new
+shop.sweet_count
+#=> 4000003
+shop.sweet_count
+#=> 4000003 # this one will be super quick!
```
Comment (andycroll) [inline _posts/2021/2021-01-25-memoize-expensive-code.md:56]:
```suggestion
#=> 4000003 # this one will be super quick!
shop.average_sweets_per_jar
#=> 8000 # this uses the cached sweet_count
```

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+layout: article
+category: ruby
+image:
+  base: '2021/memoize-expensive-code'
+  alt: "Red Floppy Disk"
+  credit:  Fredy Jacob
+  source: "https://unsplash.com/photos/t0SlmanfFcg"
+---
+
+Memoization is a performance optimization where the result of a slow or non-performant piece of code is temporarily stored and when the expensive code is called again, the stored value is returned.
```
Comment (nodunayo) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]: Comma after 'stored', I think.

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+It is a specific form of caching that is relatively elegant to read and implement in Ruby due to the use of the `||=` (or equals) operator.
```
Comment (nodunayo) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]: I would take out showing the implementation here. Save it for the big reveal! This is the intro after all. :-)

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+  end
+end
+```
+
+
+## Instead...
+
+```ruby
+class OldTimeySweetShop
+  # leave the same
```
Comment (nodunayo) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]: Maybe at least leave the `def average..` part? as `# leave the same` is a bit confusing on its own.

File: _posts/2021/2021-01-25-memoize-expensive-code.md [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+shop.sweet_count
+#=> 4000003 # this one will be super quick!
+shop.average_sweets_per_jar
+#=> 8000 # this uses the cached sweet_count
+```
+
+
+## Why?
+
+Primarily this is a performance optimization that has only a small impact on the readability of your code.
```
Comment (nodunayo) [inline _posts/2021/2021-01-25-memoize-expensive-code.md]: Maybe something like: "Given this change only has a small impact on...it's worth it for the performance optimisaiton.."

And I think maybe make this the second paragraph in this section, starting with the "In typical Rails applications.." part...

**Review by nodunayo — APPROVED** [review]

**Pull request merged**